aws/resource_aws_cloudsearch_service.go

```diff
@@ -0,0 +1,471 @@
+
+	return err
+}
+
+func resourceAwsCloudSearchDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cloudsearchconn
+
+	err := updateScalingParameters(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	updated, err := defineIndexFields(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	err = updateAccessPolicy(d, meta, conn)
+	if err != nil {
+		return err
+	}
+
+	if updated {
+		_, err := conn.IndexDocuments(&cloudsearch.IndexDocumentsInput{
+			DomainName: aws.String(d.Get("domain_name").(string)),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateScalingParameters(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) error {
+	input := cloudsearch.UpdateScalingParametersInput{
+		DomainName: aws.String(d.Get("domain_name").(string)),
+		ScalingParameters: &cloudsearch.ScalingParameters{
+			DesiredInstanceType:     aws.String(d.Get("instance_type").(string)),
+			DesiredPartitionCount:   aws.Int64(int64(d.Get("partition_count").(int))),
+			DesiredReplicationCount: aws.Int64(int64(d.Get("replication_count").(int))),
+		},
+	}
+
+	_, err := conn.UpdateScalingParameters(&input)
+
+	return err
+}
+
+func defineIndexFields(d *schema.ResourceData, meta interface{}, conn *cloudsearch.CloudSearch) (bool, error) {
+	if d.HasChange("indexes") {
+		old := make(map[string]interface{})
+		new := make(map[string]interface{})
+
+		o, n := d.GetChange("indexes")
+
+		for _, ot := range o.([]interface{}) {
+			os := ot.(map[string]interface{})
+			old[os["name"].(string)] = os
+		}
+
+		for _, nt := range n.([]interface{}) {
+			ns := nt.(map[string]interface{})
+			new[ns["name"].(string)] = ns
+		}
+
+		// Handle Removal
+		for k := range old {
+			if _, ok := new[k]; !ok {
+				deleteIndexField(d.Get("domain_name").(string), k, conn)
+			}
+		}
+
+		for _, v := range new {
+			// Handle replaces & additions
+			err := defineIndexField(d.Get("domain_name").(string), v.(map[string]interface{}), conn)
+			if err != nil {
+				return true, err
+			}
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func defineIndexField(domainName string, index map[string]interface{}, conn *cloudsearch.CloudSearch) error {
+	i, err := genIndexFieldInput(index)
+	if err != nil {
+		return err
+	}
+
+	input := cloudsearch.DefineIndexFieldInput{
+		DomainName: aws.String(domainName),
+		IndexField: i,
+	}
+
+	_, err = conn.DefineIndexField(&input)
+	return err
+}
+
+func deleteIndexField(domainName string, indexName string, conn *cloudsearch.CloudSearch) error {
+	input := cloudsearch.DeleteIndexFieldInput{
+		DomainName:     aws.String(domainName),
+		IndexFieldName: aws.String(indexName),
+	}
+
+	_, err := conn.DeleteIndexField(&input)
+	return err
+}
+
+var parseError = func(d string, t string) error {
+	return fmt.Errorf("can't convert default_value '%s' of type '%s' to int", d, t)
+}
+
+func getOk(index map[string]interface{}, prop string, t interface{}) error {
```

Comment:
Could we reword this method? I do not feel comfortable using this, knowing a GetOk method exists on the schema :/
It could be nice also to add a comment here, as it is not obvious about what it does at the moment :/

Reply:
You are absolutely right.